src/main/scala/model/TodoServer.scala

```diff
@@ -57,33 +72,92 @@ object TodoServer {
 class AddTaskListener(server: TodoServer) extends DataListener[String] {
 
   override def onData(socket: SocketIOClient, taskJSON: String, ackRequest: AckRequest): Unit = {
+//    println(taskJSON)
     val task: JsValue = Json.parse(taskJSON)
     val title: String = (task \ "title").as[String]
     val description: String = (task \ "description").as[String]
+    val username: String = server.socketToUsername(socket)
+    server.database.addTask(Task(title, description, username))
+    socket.sendEvent("all_tasks", server.tasksJSON(server.socketToUsername(socket)))
 
-    server.database.addTask(Task(title, description))
-    server.server.getBroadcastOperations.sendEvent("all_tasks", server.tasksJSON())
   }
 
 }
 
+class UpdateTaskListener(server: TodoServer) extends DataListener[Nothing] {
+  override def onData(client: SocketIOClient, data: Nothing, ackSender: AckRequest): Unit = {
+    client.sendEvent("all_tasks", server.tasksJSON(server.socketToUsername(client)))
+  }
+}
 
 class CompleteTaskListener(server: TodoServer) extends DataListener[String] {
 
   override def onData(socket: SocketIOClient, taskId: String, ackRequest: AckRequest): Unit = {
     server.database.completeTask(taskId)
-    server.server.getBroadcastOperations.sendEvent("all_tasks", server.tasksJSON())
+    socket.sendEvent("all_tasks", server.tasksJSON(server.socketToUsername(socket)))
   }
+}
 
+class RegisterUserListener(server: TodoServer) extends DataListener[String] {
+  override def onData(client: SocketIOClient, username: String, ackSender: AckRequest): Unit = {
+    if (!server.usernameToSocket.keys.toList.contains(username)) {
+      server.usernameToSocket += username -> client
+      server.socketToUsername += client -> username
+      client.sendEvent("registered", username)
+    }
+    else {
+      client.sendEvent("already_registered", username)
+      server.usernameToSocket += username -> client
+      server.socketToUsername += client -> username
+    }
+  }
+}
+
+class ShareMessageListener(server: TodoServer) extends DataListener[String] {
+  override def onData(socket: SocketIOClient, data: String, ackSender: AckRequest): Unit = {
+    val user: SocketIOClient = server.usernameToSocket(data)
+
+    val tasks = server.database.getTasks
+
+    var shared: List[String] = List()
+
+    for (task <- tasks) {
+      task.display()
+      if (task.user == data) {
+        shared = shared :+ task.id
+        println(shared)
+      }
+    }
+
+    for (task <- tasks) {
+      task.display()
+      if (task.user == server.socketToUsername(socket)) {
+        if (!shared.contains(task.id)) {
+          val currentTask = new Task(task.title, task.description, task.id, data)
+          currentTask.display()
+          server.database.addTask(currentTask)
+        }
+      }
+    }
+
+    println("send to: " + data)
+    println("from: " + server.socketToUsername(socket))
+
+    user.sendEvent("all_tasks", server.tasksJSON(server.socketToUsername(user)))
+  }
```

Comment:
adds share functionality by creating a new event and sending that event to the different users. Checks for duplicates such that when all messages from a list are shared with one another, if they were previously shared they won't re-share (no duplicate sharing)